### PR TITLE
Add browser notification

### PR DESCRIPTION
**Pull request by lungsin**
Implements #292

- When an announcement or clarification is made, a browser notification appears in addition to an alert toast. This is tested on Chrome and Safari.

image
image

Notification Flow: 
1. Upon opening TLX, notification permission is asked to the user.
2. If permission is given, whenever an announcement/clarification is made, a browser notification will appear. 

Side effect and known issue:
1. ~If the user opens multiple TLX tabs, he/she will receive multiple notifications (not necessarily simultaneous)~. Multiple tabs announcement is resolved by using notification tag, a timestamp rounding down to the nearest minute is used.
2. For some reason, the notification doesn't have API for sound (nor does it produce any sound during testing). A workaround by using HTML5 Audio was tried but the sound won't be produced until the user interacts with the webpage.

### CODE REVIEW

**Comment by fushar** [comment]
This PR is failing lint, please make sure to run Prettier

**Comment by lungsin** [comment]
@fushar seems like some of the unrelated tests fail to run. All of the failed tests complained that they can't find `routes/fixtures/state`.

**Comment by fushar** [comment]
Sorry2, will fix tonight 🤦

**Comment by prabowo02** [comment]
Commonly, a user will open many tabs at once (one problem per tab). I think this may unnecessarily annoy the user when receiving multiple notifications.

**Comment by lungsin** [comment]
I agree with that, I will try to find a way to centralize the notification.

Not sure about this, do you think it's a good idea to centralize the toast as well?

**Comment by lungsin** [comment]
Multiple tabs issue is resolved by using a notification tag, a timestamp rounding down to the nearest minute is used.

TLX

**Comment by fushar** [comment]
Please rebase ya, have fixed build error in latest master.

File: judgels-frontends/raphael/src/modules/notification/notification.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import logo from '../../assets/images/logo-header.png';
+
+export function showDesktopNotification(title, tag, message) {
+  new Notification(title, { body: message, icon: logo, tag: tag });
```
Comment (fushar) [inline judgels-frontends/raphael/src/modules/notification/notification.js]:
```suggestion
  new Notification(title, { body: `${APP_CONFIG.name}: message`, icon: logo, tag: tag });
```

File: judgels-frontends/raphael/src/routes/contests/contests/single/announcements/modules/contestAnnouncementActions.js [hunk 2]
```diff
@@ -9,9 +10,11 @@ export function getAnnouncements(contestJid, page) {
   };
 }
 
-export function alertNewAnnouncements() {
+export function alertNewAnnouncements(notificationTag) {
   return async () => {
-    toastActions.showAlertToast('You have new announcement(s).');
+    const message = 'You have new announcement(s).';
+    toastActions.showAlertToast(message);
+    showDesktopNotification('TLX New Announcement(s)', notificationTag, message);
```
Comment (fushar) [inline judgels-frontends/raphael/src/routes/contests/contests/single/announcements/modules/contestAnnouncementActions.js]:
```suggestion
    showDesktopNotification('New announcement(s)', notificationTag, message);
```

File: judgels-frontends/raphael/src/routes/contests/contests/single/clarifications/modules/contestClarificationActions.js [hunk 3]
```diff
@@ -37,12 +38,15 @@ export function answerClarification(contestJid, clarificationJid, answer) {
...
+    let message;
     if (status === ContestClarificationStatus.Answered) {
-      toastActions.showAlertToast('You have new answered clarification(s).');
+      message = 'You have new answered clarification(s).';
     } else {
-      toastActions.showAlertToast('You have new clarification(s).');
+      message = 'You have new clarification(s).';
     }
+    toastActions.showAlertToast(message);
+    showDesktopNotification('TLX Clarification', notificationTag, message);
```
Comment (fushar) [inline judgels-frontends/raphael/src/routes/contests/contests/single/clarifications/modules/contestClarificationActions.js]:
```suggestion
    showDesktopNotification('New clarification(s)', notificationTag, message);
```

Should we also distinguish between answered and new clarifications here?

File: judgels-frontends/raphael/src/routes/contests/contests/single/components/ContestAnnouncementsWidget/ContestAnnouncementsWidget.jsx [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import { Tag } from '@blueprintjs/core';
 import * as React from 'react';
 import { connect } from 'react-redux';
+import { selectContest } from '../../../modules/contestSelectors';
```
Comment (fushar) [inline judgels-frontends/raphael/src/routes/contests/contests/single/components/ContestAnnouncementsWidget/ContestAnnouncementsWidget.jsx]: Please group this with other `{ select*** }` imports

File: judgels-frontends/raphael/src/components/Header/Header.jsx [hunk 5]
```diff
@@ -8,11 +8,12 @@ import UserWidget from '../UserWidget/UserWidget';
 import './Header.css';
 
 import logo from '../../assets/images/logo-header.png';
+import { askDesktopNotificationPermission } from '../../modules/notification/notification';
```
Comment (fushar) [inline judgels-frontends/raphael/src/components/Header/Header.jsx]: Can we put this after line 6? (Still with blank space before CSS import)

File: judgels-frontends/raphael/src/components/Header/Header.jsx [hunk 6]
```diff
@@ -8,11 +8,12 @@ import UserWidget from '../UserWidget/UserWidget';
...
 
 class Header extends React.PureComponent {
   render() {
     const UW = this.props.userWidget;
-
+    askDesktopNotificationPermission();
```
Comment (fushar) [inline judgels-frontends/raphael/src/components/Header/Header.jsx]: Shouldn't this be in `componentDidMount()`?

**Review by fushar — CHANGES_REQUESTED** [review]

File: judgels-frontends/raphael/src/routes/contests/contests/single/clarifications/modules/contestClarificationActions.js (hunk 3 above)
Comment (lungsin) [reply judgels-frontends/raphael/src/routes/contests/contests/single/clarifications/modules/contestClarificationActions.js]: Hmmm yes, I think that's a good idea.

File: judgels-frontends/raphael/src/routes/contests/contests/single/components/ContestAnnouncementsWidget/ContestAnnouncementsWidget.jsx (hunk 4 above)
Comment (lungsin) [reply judgels-frontends/raphael/src/routes/contests/contests/single/components/ContestAnnouncementsWidget/ContestAnnouncementsWidget.jsx]: I'm not sure what you mean, do you mean something like this? 

image

File: judgels-frontends/raphael/src/routes/contests/contests/single/clarifications/modules/contestClarificationActions.js (hunk 3 above)
Comment (lungsin) [reply judgels-frontends/raphael/src/routes/contests/contests/single/clarifications/modules/contestClarificationActions.js]: I distinguish it by adding 'Answered' to the string:
- "New answered clarification(s)"
- "New clarification(s)"

File: judgels-frontends/raphael/src/routes/contests/contests/single/components/ContestAnnouncementsWidget/ContestAnnouncementsWidget.jsx (hunk 4 above)
Comment (fushar) [reply judgels-frontends/raphael/src/routes/contests/contests/single/components/ContestAnnouncementsWidget/ContestAnnouncementsWidget.jsx]: Yes

Comment (fushar) [reply judgels-frontends/raphael/src/routes/contests/contests/single/components/ContestAnnouncementsWidget/ContestAnnouncementsWidget.jsx]: Sorry2 what I meant was

```
import { connect } from 'react-redux';

import { selectContest } from '../../../modules/contestSelectors';
import { selectContestWebConfig } from '../../../modules/contestWebConfigSelectors';
```

File: judgels-frontends/raphael/src/components/Header/Header.jsx [hunk 7]
```diff
@@ -32,6 +33,10 @@ class Header extends React.PureComponent {
       </Navbar>
     );
   }
+
+  componentDidMount() {
```
Comment (fushar) [inline judgels-frontends/raphael/src/components/Header/Header.jsx]: Nit: can we put this before `render()`; this is the convention in React

File: judgels-frontends/raphael/src/routes/contests/contests/single/components/ContestAnnouncementsWidget/ContestAnnouncementsWidget.jsx [hunk 8]
```diff
@@ -15,12 +17,16 @@ class ContestAnnouncementsWidget extends React.Component {
 
   componentDidUpdate(prevProps) {
     if (this.props.announcementCount > prevProps.announcementCount) {
-      this.props.onAlertNewAnnouncements();
+      // TODO(lungsin): change the notification tag to be more proper, e.g. using announcement JID.
+      const timestamp = Math.floor(Date.now() / SingleContestDataRoute.GET_CONFIG_TIMEOUT); // Use timestamp for notification tag
```
Comment (fushar) [inline judgels-frontends/raphael/src/routes/contests/contests/single/components/ContestAnnouncementsWidget/ContestAnnouncementsWidget.jsx]: Can we instead add the constant in `src/modules/api/uriel/contestWeb.js` and have all 3 files refer to the const

**Review by fushar — CHANGES_REQUESTED** [review]

**Review by fushar — APPROVED** [review]

**Pull request merged**